presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java

```diff
@@ -326,6 +306,177 @@ public void testNonPushedDownJoinFilterRemoval()
+    }
+
+    @Test
+    public void testMultiSemiJoin()
+    {
+        assertPlan(
+                "SELECT part.partkey FROM part WHERE part.partkey IN (SELECT lineitem.partkey FROM lineitem WHERE lineitem.orderkey IN (SELECT orders.orderkey FROM orders))",
+                anyTree(
+                        filter(
+                                "S0",
+                                project(
+                                        semiJoin(
+                                                "PART_PK",
+                                                "LINEITEM_PK",
+                                                "S0",
+                                                true,
+                                                anyTree(
+                                                        tableScan(
+                                                                "part",
+                                                                ImmutableMap.of("PART_PK", "partkey"))),
+                                                anyTree(
+                                                        filter(
+                                                                "S1",
+                                                                project(
+                                                                        semiJoin(
+                                                                                "LINEITEM_OK",
+                                                                                "ORDERS_OK",
+                                                                                "S1",
+                                                                                true,
+                                                                                anyTree(
+                                                                                        tableScan(
+                                                                                                "lineitem",
+                                                                                                ImmutableMap.of("LINEITEM_PK", "partkey", "LINEITEM_OK", "orderkey"))),
+                                                                                anyTree(
+                                                                                        tableScan(
+                                                                                                "orders",
+                                                                                                ImmutableMap.of("ORDERS_OK", "orderkey"))))))))))));
+    }
+
+    @Test
+    public void testSemiJoinUnsupportedDynamicFilterRemoval()
+    {
+        // Dynamic filters are supported only after a table scan
+        assertPlan(
+                "WITH t AS (SELECT lineitem.partkey + 1000 partkey FROM lineitem WHERE lineitem.orderkey IN (SELECT orders.orderkey FROM orders)) " +
```

Comment:
With `WITH t AS (SELECT lineitem.partkey + 1000 partkey FROM lineitem)`, the project will be merged with table scan but we don't have `TableScanAndProjectMatcher` to represent it. Would it be good to keep the inner semi-join here as it can also serve as a comparison to the outer semi-join? We can see clearly that the inner semi-join has dynamic filter but not the outer one.

Reply:
I think
```
SELECT partkey FROM (SELECT lineitem.partkey + 1000 partkey FROM lineitem) WHERE partkey IN (SELECT part.partkey FROM part)
```
should be sufficient.
It would be easier for reader to undestand the test intent and matched tree.